### ISSUE TITLE
Support unbundles to improve app startup perf

### ISSUE THREAD

**Issue by rigdern** [Microsoft/react-native-windows#952]
Unbundles enable you to load your JavaScript code lazily 1 module/file at a time rather than having to inject the entire bundle file into the JavaScript engine during startup. One of its main benefits is that it improves startup times by only loading the minimal amount of code necessary during startup.

It's supported by the iOS and Android implementations of React Native. Unfortunately, it's currently undocumented: facebook/react-native#10838

**Comment by rozele** [comment]
This actually seems pretty easy to do.  I think we just need to add a `nativeRequire` callback and a mechanism to check if the source URL path has a file that matches the magic "UNBUNDLE" name.

**Comment by rozele** [comment]
@rigdern, if you get a chance to test this to see if it has an impact on your projects startup time, it would be great to hear.  I think the extra file IO overhead overcomes any potential savings of JS evaluation for the small projects I've tried this on.

**Comment by rigdern** [comment]
@rozele I'll see if I can find time to try it in the next couple of days.

I skimmed over your implementation and it looks like each module is contained in a separate file. From what I remember, the iOS implementation of unbundle is one big file. It opens that one file and reads portions of it as necessary. Did you consider that implementation option and how the perf would compare? I think they use a memory-mapped file although I'm not sure if UWP has APIs for that.

**Comment by rozele** [comment]
Thanks @rigdern, I started looking into the indexed-unbundle but haven't got around to implementing it.  I think we have a similar limitation that iOS has in that the increased file IO at app start impacts performance more than loading on demand.  I don't believe there is a memory-mapped file API for C#, but there is one that we were using for the C++/CX bridge implementation we have for serialized (pre-parsed) JavaScript bundles, so that's always an option.

**Comment by rozele** [comment]
@rigdern I added the indexed unbundle logic for the C#-based ChakraJavaScriptExecutor, but as already discussed above, it does not use a memory-mapped file, so the IO may not be much better than reading individual files.  I'm going to start working on porting this implementation to the C++/CX JavaScriptExecutor implementation, which will have memory-mapping.

**Comment by rozele** [comment]
@rigdern I finished the work to add indexed unbundle to the C++/CX JavaScriptExecutor.  Even for the small test app, the results are significant.

I created an app that started a stopwatch at app start and included a native module to read from that stopwatch. I read from that stopwatch from the `componentDidMount` callback in the main app component. I ran each variant 15 times to get a reasonable average (we can do much more rigorous testing later).  We have two IJavaScriptExecutor implementations, the C# version that uses P/Invoke against Chakra.dll (ChakraJavaScriptExecutor) and the C++/CX version (NativeJavaScriptExecutor).

```
ChakraJavaScriptExecutor + Regular Bundle - 291 milliseconds
NativeJavaScriptExecutor + RegularBundle - 286 milliseconds
ChakraJavaScriptExecutor + Regular Unbundle - 1369 milliseconds
NativeJavaScriptExecutor + Regular Unbundle - 257 milliseconds
ChakraJavaScriptExecutor + Indexed Unbundle - 257 milliseconds
NativeJavaScriptExecutor + Indexed Unbundle - 214 milliseconds
```

So the elapsed milliseconds on the stopwatch is roughly representative of the overhead to start the app with each bundle type. The unbundle using the managed JavaScriptExecutor has horrendous performance because of the overhead of the UWP filesystem APIs.  The indexed unbundle on the C++/CX JavaScriptExecutor is the fastest. It uses a memory mapped file as well.
